### PR TITLE
TLSv1.3: serverinfo support

### PR DESCRIPTION
**Pull request by mattcaswell**
We currently provide the SSL_CTX_use_serverinfo() and int SSL_CTX_use_serverinfo_file() functions which provide an even simpler interface to the custom extensions API which can also be configured at runtime, i.e. at runtime you can configure applications to read a serverinfo file which contains details of custom extensions to be added to the ServerHello if the corresponding extension was received in the ClientHello. For example this can be used to add Certificate Transparency details to servers. We only provide built-in support for CT on the client side - to do the server side you would typically use custom extensions or the serverinfo mechanism.

However the servinfo mechanism was built on top of the old TLSv1.2 and below custom extensions API. The file format provides no way of specifying whether extensions should be added to the TLSv1.3 ServerHello, EncryptedExtensions, Certificate etc.

This PR provides a new serverinfo file format and an SSL_CTX_use_serverinfo_ex() function for reading new (or old) format serverinfo data. SSL_CTX_use_serverinfo_file() can dynamically work out what file format (old or new) has been provided - and does the right thing.

##### Checklist

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Comment by snhenson** [comment]
Are there any circumstances where we might want to attach extensions to certificates that aren't the first?

**Comment by richsalz** [comment]
I think the major use-case is adding things to the leaf/end-entity cert, and only that.  Things like OCSP or CT stapled data for example.

File: doc/man3/SSL_CTX_use_serverinfo.pod [hunk 1]
```diff
@@ -2,33 +2,59 @@
...
 
  int SSL_CTX_use_serverinfo_file(SSL_CTX *ctx, const char *file);
 
 =head1 DESCRIPTION
 
-These functions load "serverinfo" TLS ServerHello Extensions into the SSL_CTX.
-A "serverinfo" extension is returned in response to an empty ClientHello
+These functions load "serverinfo" TLS extensions into the SSL_CTX. A
+"serverinfo" extension is returned in response to an empty ClientHello
 Extension.
```
Comment (richsalz) [inline doc/man3/SSL_CTX_use_serverinfo.pod]: ", and is only used in TLS 1.3" ?

File: doc/man3/SSL_CTX_use_serverinfo.pod [hunk 2]
```diff
@@ -2,33 +2,59 @@
...
+SSL_CTX_use_serverinfo_ex() loads one or more serverinfo extensions from
+a byte array into B<ctx>. The B<version> parameter specifies the format of the
+byte array provided in B<*serverinfo> which is of length B<serverinfo_length>.
+
+If B<version> is B<SSL_SERVERINFOV2> then the extensions in the array must
+consist of a 4-byte context, a 2-byte Extension Type, a 2-byte length, and then
+length bytes of extension_data. The context and type values have the same
+meaning as for L<SSL_CTX_add_custom_ext(3)>. If serverinfo is being loaded for
+extensions to be added to a Certificate message, then the extension will only
+be added for the first Certificate in the message.
```
Comment (richsalz) [inline doc/man3/SSL_CTX_use_serverinfo.pod]: is it the FIRST or is it the leaf?  :)  Consider adding ", which is end-entity certificate" or some such.

File: doc/man3/SSL_CTX_use_serverinfo.pod [hunk 3]
```diff
@@ -2,33 +2,59 @@
...
 SSL_CTX_use_serverinfo_file() loads one or more serverinfo extensions from
 B<file> into B<ctx>.  The extensions must be in PEM format.  Each extension
-must consist of a 2-byte Extension Type, a 2-byte length, and then length
-bytes of extension_data.  Each PEM extension name must begin with the phrase
-"BEGIN SERVERINFO FOR ".
+must be in a format as described above for SSL_CTX_use_serverinfo_ex().  Each
+PEM extension name must begin with the phrase "BEGIN SERVERINFOV2 FOR " for
+SSL_SERVERINFOV2 data or "BEGIN SERVERINFO FOR " for SSL_SERVERINFOV1 data.
 
 If more than one certificate (RSA/DSA) is installed using
```
Comment (richsalz) [inline doc/man3/SSL_CTX_use_serverinfo.pod]: is this paragraph still accurate?
If so, it seems to contradict the paragraph starting at line 33.

File: doc/man3/SSL_CTX_use_serverinfo.pod (hunk 1 above)
Comment (mattcaswell) [inline doc/man3/SSL_CTX_use_serverinfo.pod]: No - this applies to all protocol versions. "serverinfo" is an existing capability which we're just upgrading to be able to cope with TLS1.3 too.

File: doc/man3/SSL_CTX_use_serverinfo.pod (hunk 3 above)
Comment (mattcaswell) [inline doc/man3/SSL_CTX_use_serverinfo.pod]: Does it? How so? SSL_CTX_use_serverinfo_ex() works with binary data, while SSL_CTX_use_serverinfo_file() works with PEM format data and uses the PEM header to determine what format binary data is contained within it.

Comment (richsalz) [inline doc/man3/SSL_CTX_use_serverinfo.pod]: the first para talks about only applying to the FIRST certificate.  The para at line 59 talks about "if more than one"  Maybe I just misread it.

Comment (mattcaswell) [inline doc/man3/SSL_CTX_use_serverinfo.pod]: Oh, I see. There is a distinction between certificates loaded into the SSL_CTX (e.g. for a server we are using an RSA certificate or an ECDSA certificate (or both)) and the certificates returned as part of the Certificate message (which start with the EE cert selected and then provide the rest of the chain). For example if you have configured you server for both RSA and ECDSA you could have different serverinfo data associated with each one. We only actually add the extension to the EE cert (i.e. the first one in the Certificate message).

Comment (richsalz) [inline doc/man3/SSL_CTX_use_serverinfo.pod]: ah, got it.  So just tweak "each time a certificate is loaded." to be "each call to ```SSL_CTX_use_certificate``` ?
+1 either way.

File: doc/man3/SSL_CTX_use_serverinfo.pod (hunk 2 above)
Comment (mattcaswell) [inline doc/man3/SSL_CTX_use_serverinfo.pod]: Done.

File: doc/man3/SSL_CTX_use_serverinfo.pod (hunk 3 above)
Comment (mattcaswell) [inline doc/man3/SSL_CTX_use_serverinfo.pod]: I changed it to "...after each time a certificate is loaded via a call to SSL_CTX_use_certificate()"

**Comment by mattcaswell** [comment]
Pushed. Thanks.

**Pull request closed without merging**